libsolidity/analysis/TypeChecker.cpp

```diff
@@ -3859,6 +3986,95 @@ void TypeChecker::endVisit(UsingForDirective const& _usingFor)
 					result.message().empty() ? "." : ": " +  result.message()
 				)
 			);
+		else if (operator_)
+		{
+			TypePointers const& parameterTypes = functionType->parameterTypesIncludingSelf();
+			size_t const parameterCount = parameterTypes.size();
+			if (usingForType->category() != Type::Category::UserDefinedValueType)
+			{
+				m_errorReporter.typeError(
+					5332_error,
+					path->location(),
+					"Operators can only be implemented for user-defined value types."
+				);
+				continue;
+			}
+			solAssert(usingForType->typeDefinition());
+
+			if (functionType->stateMutability() != StateMutability::Pure)
+				m_errorReporter.typeError(
+					7775_error,
+					path->location(),
+					"Only pure functions can be used to define operators."
+				);
+
+			bool identicalFirstTwoParameters = (parameterCount < 2 || *parameterTypes.at(0) == *parameterTypes.at(1));
```

Comment:
If I was @cameel, I might complain about something that has none or only one parameter not really having identical first two parameters ;-).

Reply:
That's because you're looking at it in absolute terms :) I think it's still a big relative improvement over the [monster conditions in the old PR](https://github.com/ethereum/solidity/pull/12362/files#diff-1286fd2f351e6a40b993189fde18c5db99ddfe4a35f4ada894945ba62819a324R3965-R3986) :)

Reply:
Also, when you have less than two parameters, it's not the case where the first two parameters don't match, so I find this way of formulating it quite intuitive.

Reply:
Arguing as a Russellian, it's true, that it's false that the first two parameters don't match in this case, but it's also true that it's false that they do match ;-P.

Reply:
But yeah, the checking here generally seems a bit convoluted, but we can also keep it.

Reply:
Yeah, there's just a lot going on in these conditions and I did not find a way to do it better. This seemed like a good compromise between being precise and keeping it readable.